dev/src/audio.cpp

```diff
@@ -22,20 +22,72 @@ using namespace lobster;
 
 void AddSound(NativeRegistry &nfr) {
 
-nfr("play_wav", "filename,volume", "SI?", "B",
+nfr("play_wav", "filename,loops", "SI?", "I",
     "plays a sound defined by a wav file (RAW or MS-ADPCM, any bitrate other than 22050hz 16bit"
-    " will automatically be converted on first load). volume in range 1..128, or omit for max."
-    " returns false on error",
-    [](StackPtr &, VM &, Value &ins, Value &vol) {
-        bool ok = SDLPlaySound(ins.sval()->strv(), false, vol.True() ? vol.intval() : 128);
-        return Value(ok);
+    " will automatically be converted on first load). the default volume is the max volume (1.0)"
+    " loops is the number of repeats to play (-1 repeats endlessly, omit for no repeats)."
+    " returns the assigned channel number (1..8) or 0 on error",
+    [](StackPtr &, VM &, Value &ins, Value &loops) {
+        int ch = SDLPlaySound(ins.sval()->strv(), false, 1.0, loops.True() ? loops.intval() : 0);
+        return Value(ch);
     });
 
-nfr("play_sfxr", "filename,volume", "SI?", "B",
+nfr("play_sfxr", "filename,loops", "SI?", "I",
     "plays a synth sound defined by a .sfs file (use http://www.drpetter.se/project_sfxr.html"
-    " to generate these). volume in range 1..128, or omit for max. returns false on error",
-    [](StackPtr &, VM &, Value &ins, Value &vol) {
-        bool ok = SDLPlaySound(ins.sval()->strv(), true, vol.True() ? vol.intval() : 128);
+    " to generate these). the default volume is the max volume (1.0)"
+    " loops is the number of repeats to play (-1 repeats endlessly, omit for no repeats)."
+    " returns the assigned channel number (1..8) or 0 on error",
+    [](StackPtr &, VM &, Value &ins, Value &loops) {
+        int ch = SDLPlaySound(ins.sval()->strv(), true, 1.0, loops.True() ? loops.intval() : 0);
+        return Value(ch);
+    });
+
+nfr("sound_halt", "channel", "I", "B",
+    "terminates sound channel. returns false on error",
+    [](StackPtr &, VM &, Value &ch) {
+        bool ok = SDLHaltSound(ch.True() ? ch.intval() : -1);
+        return Value(ok);
+    });
+
+nfr("sound_pause", "channel", "I", "B",
+    "pauses sound channel. returns false on error",
+    [](StackPtr &, VM &, Value &ch) {
+        int ch_idx = ch.intval();
+        bool ok = false;
+        if (ch_idx != -1)
```

Comment:
Don't think its that useful for any functions other than `play_` ones to return an error, they should simply silently do nothing on wrong channel. And init is not going to fail on these functions since they're never the first ones to be called

Reply:
"And init is not going to fail on these functions since they're never the first ones to be called" -- unless of course a Lobster user is calling them before calling any play. However, the behavior of the corresponding Mix_xxx functions in mixer.c is indeed benign I think so yes, it won't crash anything.